app/uk/gov/hmrc/gform/controllers/AuthenticatedRequestActions.scala

```diff
@@ -400,24 +457,32 @@ class AuthenticatedRequestActions(
 
         formModelOptics <-
           FormModelOptics
-            .mkFormModelOptics[DataOrigin.Mongo, Future, U](cache.variadicFormData, cache, recalculation)
+            .mkFormModelOptics[DataOrigin.Mongo, Future, U](
+              cache.variadicFormData,
+              cache,
+              recalculation
+            )
 
-        formModelOpticsUpd =
-          formModelOptics
-            .modify(_.formModelVisibilityOptics.recalculationResult.evaluationContext.formTemplateId)
-            .setTo(formUpd.formTemplateId)
+        (cacheUpdated, formModelOpticsUpdated) <-
+          maybeUpdateItmpCacheAndOptics(
```

Comment:
This is executed on every request. As discussed before, we should do it only once when user enters the form and so it should be implemented in `NewFormController.scala`